packages/protocol/src/utils/SPID.ts

```diff
@@ -0,0 +1,202 @@
+     */
+    equals(other: SPIDLike): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = SPID.from(other)
+        } catch (_err) {
+            // ignore error
+            // not equal if not valid
+            return false
+        }
+
+        // check key matches
+        return this.key === otherSpid.key
+    }
+
+    /**
+     * Convert SPIDLikePartial to SPIDObjectPartial
+     * i.e. normalizes various input types/shapes to { streamId?, streamPartition? }
+     * Note: does not throw on malformed input
+     */
+    static toSPIDObjectPartial(spidLike: SPIDLikePartial): SPIDObjectPartial {
+        // convert from string
+        if (typeof spidLike === 'string') {
+            const [streamId, partitionStr] = spidLike.split(this.SEPARATOR)
+            // partition is optional, falls back to undefined i.e. default
+            const streamPartition = partitionStr != null ? Number.parseFloat(partitionStr) : undefined
+            return { streamId, streamPartition }
+        } else if (spidLike && typeof spidLike === 'object') {
+            // @ts-expect-error object should have one of these, validated anyway
+            const streamId = spidLike.streamId || spidLike.id
+            // @ts-expect-error object should have one of these, validated anyway
+            const partition = spidLike.streamPartition || spidLike.partition
+            // try parse if a value was passed, but fall back to undefined i.e. default
+            const streamPartition = partition != null ? Number.parseFloat(partition) : undefined
+            return { streamId, streamPartition }
+        } else {
+            return { streamId: undefined, streamPartition: undefined }
+        }
+    }
+
+    /**
+     * Convert to SPID if possible, with defaults.
+     * e.g.
+     * ```ts
+     * fromDefaults(streamId, { partition: 0 })
+     * ```
+     */
+    static fromDefaults(spidLike: SPIDLike, defaultValues?: SPIDLikePartial): SPID
+    static fromDefaults(spidLike: SPIDLikePartial, defaultValues: SPIDLikePartial): SPID // requires id+partition if no defaults
+    static fromDefaults(spidLike: SPIDLikePartial, defaultValues?: SPIDLikePartial): SPID {
+        // return spid if already spid
+        if (spidLike instanceof SPID) {
+            return spidLike
+        }
+
+        // defaults can be partial, e.g. { partition: 0 }
+        // toSPIDObjectPartial can handle undefined input but we want external interface to check for it.
+        const defaults = SPID.toSPIDObjectPartial(defaultValues!)
+        const { streamId = defaults?.streamId, streamPartition = defaults?.streamPartition } = SPID.toSPIDObjectPartial(spidLike)
+        try {
+            // constructor can handle partial input but we want external interface to check for it.
+            return new SPID(streamId!, streamPartition!)
+        } catch (err) {
```

Comment:
Is this catch necessary since constructor already throw a `SPIDValidationError` if the input is not valid?

Reply:
The catch error shows the input to the `from` call.